doc/musig-spec.mediawiki

```diff
@@ -0,0 +1,222 @@
+** Let ''a<sub>i</sub> = KeyAggCoeff(pk<sub>1..u</sub>, pk<sub>i</sub>)''.
+** Let ''P<sub>i</sub> = lift_x(int(pk<sub>i</sub>))''; fail if it fails.
+* Let ''Q = a<sub>1</sub>⋅P<sub>1</sub> + a<sub>2</sub>⋅P<sub>1</sub> + ... + a<sub>u</sub>⋅P<sub>u</sub>''
+* Fail if ''is_infinite(Q)''.
+* Return ''bytes(Q)''.
+
+The algorithm ''HashKeys(pk<sub>1..u</sub>)'' is defined as:
+* Return ''hash<sub>KeyAgg list</sub>(pk<sub>1</sub> || pk<sub>2</sub> || ... || pk<sub>u</sub>)''
+
+The algorithm ''IsSecond(pk<sub>1..u</sub>, pk')'' is defined as:
+* For ''j = 1 .. u'':
+** If ''pk<sub>j</sub> &ne; pk<sub>1</sub>'':
+*** Return ''true'' if ''pk<sub>j</sub> = pk' '', otherwise return ''false''.
+* Return ''false''
+
+The algorithm ''KeyAggCoeff(pk<sub>1..u</sub>, pk)'' is defined as:
+* Let ''L = HashKeys(pk<sub>1..u</sub>)''.
+* Return 1 if ''IsSecond(pk<sub>1..u</sub>, pk<sub>i</sub>)'', otherwise return  ''int(hash<sub>KeyAgg coefficient</sub>(L || pk) mod n''.
+
+==== Nonce Generation ====
+
+The algorithm ''NonceGen()'' is defined as:
+* Generate two random integers ''k<sub>1</sub>, k<sub>2</sub>'' in the range ''1...n-1''
+* Let ''R<sup>*</sup><sub>1</sub> = k<sub>1</sub>⋅G, R<sup>*</sup><sub>2</sub> = k<sub>2</sub>⋅G''
+* Let ''pubnonce = cbytes(R<sup>*</sup><sub>1</sub>) || cbytes(R<sup>*</sup><sub>2</sub>)''
+* Let ''secnonce = bytes(k<sub>1</sub>) || bytes(k<sub>2</sub>)''
+* Return ''secnonce'' and ''pubnonce''
+
+==== Nonce Aggregation ====
+
+* The number ''u'' of ''pubnonces'' with ''0 < u < 2^32''
+* The public nonces ''pubnonce<sub>1..u</sub>'': ''u'' 66-byte arrays
+
+The algorithm ''NonceAgg(pubnonce<sub>1..u</sub>)'' is defined as:
+* For ''i = 1 .. 2'':
+** For ''j = 1 .. u'':
+*** Let ''R<sub>i,j</sub> = lift_x(int(pubnonce<sub>j</sub>[i*33:i*33+33]))''; fail if that fails
+** Let ''R'<sub>i</sub> = R<sub>i,1</sub> + R<sub>i,2</sub> + ... + R<sub>i,u</sub>''
+** Let ''R<sub>i</sub> = R'<sub>i</sub>'' if not ''is_infinite(R'<sub>i</sub>)'', otherwise let R<sub>i</sub> = G''
+* Return ''aggnonce = cbytes(R<sub>1</sub>) || cbytes(R<sub>2</sub>)''
+
+Note: If ''is_infinite(R'<sub>i</sub>)'' there is at least one dishonest signer.
+If we would fail here, we will never be able to determine who it is.
+Therefore, we should continue such that the culprit is revealed when collecting and verifying partial signatures.
+However, dealing with the point at infinity requires defining a serialization and may require extra code complexity in implementations.
+Instead, we set the aggregate nonce to some arbitrary point, the generator.
+This is secure, because it only restricts the abilities of the attacker: an attacker that forces the sum of nonces to be infinity by sending some maliciously generated nonce pairs can be turned into an attacker that forces the sum to be the generator (by simply adding the generator to one of the malicious nonces), and this does not change the winning condition of the signature security game (EUF-CMA).
```

Comment:
I don't understand. You're saying that the attacker can send `R_neg = - R_others` and `R_neg + G` and both will result in the same final nonce R. If you get two partial sigs with the same R but two different values of c, then the secnonce has been reused. The attacker can also get same R different c by choosing a different message if the secnonce is reused.

Reply:
Oops, what I said before was nonsense. I forgot that the adversary also needs the private nonce, haha.

Is it correct if an adversary A can win the EUF-CMA game against the modified scheme (call it MuSig2'), then A can be used directly to also win the EUF-CMA game against MuSig2*?

I'm thinking every query that A makes against a MuSig2' challenger (H_agg, H_non, H_sig, Sign, Sign') is also a valid query against a MuSig2* challenger (I assume this is what you meant by "only restricts the abilities of the attacker"), and a valid forgery against MuSig2' is also a valid forgery against MuSig2*; thus the "reduction" is completely trivial.

Reply:
Hm, I had to think about this quite a bit again which is an indication that we should expand on this argument a bit. Well, now we have the space for it.

> then A can be used directly to also win the EUF-CMA game against MuSig2*?

I don't think so. If A sends a Sign' query with aggnonce[0] = infinity, it expects a partial signature using aggnonce[0] = G. Therefore, a reduction would need to add G to A's query if it is infinity.

> a valid forgery (L, m, (R, s)) against MuSig2' is also a valid forgery against MuSig2*

What confused me a bit is that R = infinity is allowed in principle in MuSig2*. The scheme we're comparing this to is where R = infinity in a forgery is disallowed already (which is what BIP340 also implements).

Reply:
Ah, I think I see what you're saying.

I was previously considering`NonceAgg` as being done by the adversary, including the part where it possibly replaces infinity with the generator (because `NonceAgg` only depends on public data). In this case, I think if Sign' for MuSig2' receives infinity as one of the aggnonces, it should just fail.

But if replacing infinity with the generator is done inside Sign', by the MuSig2' challenger, then I agree with you that the reduction needs to replace infinity with the generator before passing the query onto the MuSig2* challenger.

Reply:
Here's a suggestion for an alternate way to argue about the security of the modification. I'm not 100% confident in it but if this general idea works I think it might be easier to understand:

> Note that NonceAgg (both the original and modified version) only depends on publicly available data (the set of public pre-nonces from every signer). Thus in the context of the signature security game (EUF-CMA), we can consider NonceAgg to be performed by the adversary (rather than the challenger) without loss of generality.
>
> This modification changes neither the behavior of the EUF-CMA challenger nor the condition required to win the security game (the adversary still has to output a valid forgery according to the unmodified MuSig2* scheme). Since we've already proved that MuSig2* is secure against an arbitrary adversary, we can conclude that the modified scheme is still secure.

Reply:
> But if replacing infinity with the generator is done inside Sign', by the MuSig2' challenger, then I agree with you

You're right, what I said is nonsense. Aggregation happens outside of Sign' exactly because we don't want to trust aggregation.

The sentence "it only restricts the abilities of the attacker" is wrong, because the abilities of the attacker are exactly the same (arbitrary `NonceAgg`). I pushed your suggestion (after minor modifications).

CC @real-or-random

Reply:
I was also confused first but I think now the argument makes sense. Wording can still be improved. Right now it sounds like we're coming up with this argument here in the spec but the EUF-CMA game in the paper already lets the adversary perform NonceAgg (=SignAgg in the paper).

So I believe a formal argument is as simple as this. 

Let MuSig2* be the scheme from the paper, let MuSig2† be MuSig2* with the modification that the scheme simply aborts if one of the aggregate nonces is infinity, and let MuSig2†† be the scheme in the spec. (Note that MuSig2†† also aborts if the one of the aggregate nonces is infinity because then parsing fails.)

MuSig2* is EUF-CMA by the theorem in the paper. Since Sign† only adds aborts to Sign*, and the adversary can simulate internally whether Sign† will hit those aborts (by checking whether one of the aggregate nonces is infinity), MuSig2† is EUF-CMA. Since MuSig2† and MuSig2†† differ only in NonceAgg, the EUF-CMA games of MuSig2† and MuSig2†† are identical. Thus MuSig2†† is EUF-CMA.

(This could also be written nicely in game hopping style.)

Reply:
OK, I think I'm following—my problem with my informal argument above is that it doesn't account for the extra "abort" possibility (it used without proof the fact that MuSig2† is secure). Here's my understanding of your formal argument, is this correct?

- Suppose we have adversary A†† that wins MuSig2†† game[1]; here's how we build A†:
	- Use A†† directly, since MuSig2† and MuSig2†† are the same game

- Suppose we have adversary A† that wins MuSig2† game; here's how we build A*:
	- Start with transparent wrapper around A†
	- Whenever A† makes call to Sign†, first check public nonces
		- If either of them is infinity, return "abort" to A†
		- Otherwise, forward call to Sign* and return result to A†
	- Wrapper simulates Sign† perfectly to A†, and everything else is unchanged
	- The forgery that A† outputs wins MuSig2† (of course) but it also wins MuSig2*

Thus if we have adversary A†† that wins MuSig2†† game, then we could build A* that wins MuSig2* game. From the paper, we can't have adversary A* that wins MuSig2* game. So taking the contrapositive, we can't have adversary A†† that wins MuSig2†† game. So the scheme in the spec is secure.

[1] I mean "wins ... game *with non-negligible probability*" everywhere, but I'll leave that out to keep it shorter.

Reply:
Yes, exactly correct! So your wrapper around A† is the formal implementation of my sentence "the adversary can simulate internally whether Sign† will hit those aborts (by checking whether one of the aggregate nonces is infinity)". 

I think the "abort" thing is a very minor detail here -- but I believe it's necessary in a fully rigorous proof: by observing aborts, the adversary could learn new information (say Sign† aborts iff the last bit of the secret key is 1).

But since the aborts in fact depend only on publicly available information, this is not the case here. And your reduction formalizes this intuition: whatever can be learned from a Sign† oracle (which can abort) can also be learned from a Sign* oracle (which does not abort), simply by determining locally whether Sign† would abort.

This also means that the current text is slightly wrong, when it says "This modification changes neither the behavior of the EUF-CMA challenger". The modification *does* change the behavior of the EUF-CMA challenger, namely by introducing aborts -- but it does so only in a way that anyhow can be simulated by the adversary.